incubator/orm/indexer.go

```diff
@@ -0,0 +1,119 @@
+package orm
+
+import (
+	"encoding/binary"
+	"math"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// IndexerFunc creates one or multiple index keys for the source object.
+type IndexerFunc func(value interface{}) ([][]byte, error)
+
+// Indexer manages the persistence for an index based on searchable keys and operations.
+type Indexer struct {
+	indexerFunc IndexerFunc
+	addPolicy   func(store sdk.KVStore, secondaryIndexKey []byte, rowId uint64) error
+}
+
+func NewIndexer(indexerFunc IndexerFunc, unique bool) *Indexer {
+	i := &Indexer{indexerFunc: indexerFunc}
+	if unique {
+		i.addPolicy = uniqueKeysAddPolicy
+	} else {
+		i.addPolicy = multiKeyAddPolicy
+	}
+	return i
+}
+
+func (u Indexer) OnCreate(store sdk.KVStore, rowId uint64, value interface{}) error {
+	secondaryIndexKeys, err := u.indexerFunc(value)
+	if err != nil {
+		return err
+	}
+
+	for _, secondaryIndexKey := range secondaryIndexKeys {
+		if err := u.addPolicy(store, secondaryIndexKey, rowId); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (u Indexer) OnDelete(store sdk.KVStore, rowId uint64, value interface{}) error {
+	secondaryIndexKeys, err := u.indexerFunc(value)
+	if err != nil {
+		return err
+	}
+	for _, secondaryIndexKey := range secondaryIndexKeys {
+		indexKey := makeIndexPrefixScanKey(secondaryIndexKey, rowId)
+		store.Delete(indexKey)
+	}
+	return nil
+}
+
+func (u Indexer) OnUpdate(store sdk.KVStore, rowId uint64, newValue, oldValue interface{}) error {
+	oldSecIdxKeys, err := u.indexerFunc(oldValue)
+	if err != nil {
+		return err
+	}
+	newSecIdxKeys, err := u.indexerFunc(newValue)
+	if err != nil {
+		return err
+	}
+	for _, oldIdxKey := range difference(oldSecIdxKeys, newSecIdxKeys) {
+		store.Delete(makeIndexPrefixScanKey(oldIdxKey, rowId))
+	}
+	for _, newIdxKey := range difference(newSecIdxKeys, oldSecIdxKeys) {
+		if err := u.addPolicy(store, newIdxKey, rowId); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// uniqueKeysAddPolicy enforces keys to be unique
+func uniqueKeysAddPolicy(store sdk.KVStore, secondaryIndexKey []byte, rowId uint64) error {
+	it := store.Iterator(makeIndexPrefixScanKey(secondaryIndexKey, 0), makeIndexPrefixScanKey(secondaryIndexKey, math.MaxUint64))
+	defer it.Close()
+	if it.Valid() {
+		return ErrUniqueConstraint
+	}
+
+	indexKey := makeIndexPrefixScanKey(secondaryIndexKey, rowId)
+	store.Set(indexKey, []byte{0})
```

Comment:
huh... it took me a while to fully grok this.

So if rows 1, 4, 7 all return secondaryIndexKey "bob", you will not store:
`"bob": {1, 4, 7}`
but rather:
`"bob0001": 0, "bob0004": 0, "bob0007": 0`

This is an important design decision with some interesting trade-offs (improves performance for queries and small updates I think, while making creation and large updates more expensive).

This should definitely be in the docs, and we could have a discussion with Aaron on this.  Looking at [the sdk gas pricing](https://github.com/cosmos/cosmos-sdk/blob/4ffabb65a5c07dbb7010da397535d10927d298c1/store/types/gas.go#L153-L163)

```go
// KVGasConfig returns a default gas config for KVStores.
func KVGasConfig() GasConfig {
	return GasConfig{
		HasCost:          1000,
		DeleteCost:       1000,
		ReadCostFlat:     1000,
		ReadCostPerByte:  3,
		WriteCostFlat:    2000,
		WriteCostPerByte: 30,
		IterNextCostFlat: 30,
	}
}
```

We can see that 2 writes of 1 bytes = 4060 gas
And 1 write of 69 bytes = 4070 gas

It really depends on how many items we expect for one index, but I think this would be very important design criteria. I don't think it needs many changes outside of this one file now, but let's discuss this please (we can merge first, but before using in a tagged release)

Reply:
Good point with the gas price. I would be happy to discuss the design. It was defined by @aaronc in the original branch already.  I got used to it now but it definitively comes with some learning curve.

Reply:
Understood. We definitely need input from @aaronc here. And document the design decision.

I remember one of his arguments (perhaps the main one) for fixed 64-bit rowIDs was that you had a fixed size, so you could easily store multiple entries in a secondary index lookup table. Interesting that he wrote an algorithm that didn't make use of it. The idea did make sense though.

If there are 6 rows that match a secondary index, we store 48 bytes on the lookup, all 6 rowIDs simply concatenated together.

Reply:
The general idea was to store each index entry separately like ` "bob0001": 0, "bob0004": 0, "bob0007": 0`. Fixed size row ID's make it easy to know the boundary of the index key and primary key for prefix scans and I think will generally scale well for large group sizes. But the gas costs you're pointing out possibly invalidate that if I understand correctly. Basically that any single read or write has a flat cost plus a byte cost - so that it's advantageous to group multiple entries into a single KV pair. With that in mind maybe we need to reassess. If `"bob"` appears a large number of times, `"bob": {1, 4, 7}` this approach will be hard to scale. For instance if we have an address-vote index, or any sort of index for a large group. Maybe there is a middle ground approach.

What do you mean by "Interesting that he wrote an algorithm that didn't make use of it." @ethanfrey ?

Reply:
@aaronc 's concept is implemented for the second indexes. Look for calls to `makeIndexPrefixScanKey` and `stripRowIDFromIndexKey`.